### PR TITLE
feat(output.nats): Allow disabling stream creation for externally managed streams

### LINKED ISSUE #17072: outputs.nats.jetstream Attempts to Modify Existing Stream incorrectly
### Relevant telegraf.conf

```toml
[[outputs.nats]]
  servers = ["nats://nats:4222"]
  subject = "mymetrics.events.foo.bar"
  data_format = "influx"
  name = "myhostname"
  credentials = "/etc/telegraf/nats.cred"

[outputs.nats.jetstream]
  name = "mymetrics" ## This is the name of the JetStream stream
```

### Logs from Telegraf

```text
2025-05-29T15:42:56Z I! [outputs.nats] Stream (mymetrics) successfully created or updated
```

### System info

Telegraf 1.34.4 Ubuntu 22.04

### Docker

_No response_

### Steps to reproduce

1.
2.
3.
...

### Expected behavior

The stream is managed on the NATS Cluster and is configured to contain all messages using the subject filter `mymetrics.events.>`. The plugin should use the stream as-is.

### Actual behavior

The plugin updates the stream and replaces the subject filter with its own.

### Additional info

A simple fix could be that there is param for managed streams, like:
```
[outputs.nats.jetstream]
  name = "mymetrics"
  server_managed = true
```

Possible param names:
server_managed = true
stream_autocreate = false
ensure_stream = false

### PR DESCRIPTION
**Pull request by mgale**
## Summary
Provides a new param that stops Telegraf from trying to alter a stream that is centrally managed on the nats cluster. This should be backwards compatible with existing configurations and functionality.

## Checklist
- [X] No AI generated code was used in this PR

## Related issues
https://github.com/influxdata/telegraf/issues/17072

resolves #17072

### CODE REVIEW

File: plugins/outputs/nats/sample.conf [hunk 1]
```diff
@@ -33,6 +33,12 @@
   ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md
   data_format = "influx"
 
+  ## Set to true if the stream is managed externally and Telegraf should not attempt
+  ## to create or update it. When true, Telegraf will connect to the existing stream
+  ## and fail if it does not exist.
+  ## Default is false (Telegraf may create or modify the stream).
+  # external_stream_config = false
```
Comment (srebhan) [inline plugins/outputs/nats/sample.conf]: The commented option should always be the default so we can get rid of that line
```suggestion
  # external_stream_config = false
```

File: plugins/outputs/nats/nats_test.go [hunk 2]
```diff
@@ -119,6 +178,42 @@ func TestConnectAndWriteIntegration(t *testing.T) {
...
+					Cmd:             []string{"10"},
+					Entrypoint:      []string{"sleep"},
+					HostAccessPorts: []int{srvPort},
+					Env: map[string]string{
+						"NATS_URL": fmt.Sprintf("nats://%s:%s", testcontainers.HostInternal, tc.container.Ports[natsServicePort]),
+					},
+				}
+				err = natsCli.Start()
+				require.NoError(t, err, "failed to start cli container")
+				natsCli.PrintLogs()
```
Comment (srebhan) [inline plugins/outputs/nats/nats_test.go]: Ummm you never stop the container... Furthermore, why do you need to print the logs?

File: plugins/outputs/nats/nats_test.go [hunk 3]
```diff
@@ -119,6 +178,42 @@ func TestConnectAndWriteIntegration(t *testing.T) {
...
+				srvPort, err := strconv.Atoi(hostPort)
+				require.NoError(t, err, "failed to convert port to int")
+
+				// The HostAccessPorts is not setup until after the container is started
+				// so we override the entrypoint to sleep until the container is started
+				// and then run the setup commands
+				natsCli := testutil.Container{
+					Image:           "bitnami/natscli",
+					Cmd:             []string{"10"},
+					Entrypoint:      []string{"sleep"},
```
Comment (srebhan) [inline plugins/outputs/nats/nats_test.go]: This asks for trouble or at least flaky tests... :-) Can't we somehow achieve the configuration  without that container e.g. with the nats client?

File: plugins/outputs/nats/nats.go [hunk 4]
```diff
@@ -139,11 +140,24 @@ func (n *NATS) Connect() error {
 		if err != nil {
 			return fmt.Errorf("failed to connect to jetstream: %w", err)
 		}
+
+		if n.ExternalStreamConfig {
```
Comment (srebhan) [inline plugins/outputs/nats/nats.go]: Does it make sense to move this flag to the `StreamConfig` as it will only be valid for jetstream, wouldn't it?

File: plugins/outputs/nats/nats_test.go (hunk 3 above)
Comment (mgale) [reply plugins/outputs/nats/nats_test.go]: Do you mean using the nats go client libs or the nats binary client? The above code is using the nats binary inside the client container. Unfortunately the option `HostAccessPorts` is only accessible after the container has started based on my understanding and testing. I will see if I can improve this setup somehow as it is not ideal.

File: plugins/outputs/nats/nats_test.go (hunk 2 above)
Comment (mgale) [reply plugins/outputs/nats/nats_test.go]: This section has been removed.

File: plugins/outputs/nats/nats_test.go (hunk 3 above)
Comment (mgale) [reply plugins/outputs/nats/nats_test.go]: I have updated this area and made a separate call in Go to create the stream, that removes the entire container setup.

File: plugins/outputs/nats/nats.go (hunk 4 above)
Comment (mgale) [reply plugins/outputs/nats/nats.go]: Good point, I have moved the param under the Jetstream config section.

File: plugins/outputs/nats/sample.conf [hunk 5]
```diff
@@ -66,3 +66,8 @@
     # allow_rollup_hdrs = false
     # allow_direct = true
     # mirror_direct = false
+
+    ## Set to true if the stream is managed externally and Telegraf should not attempt
+    ## to create or update it. When true, Telegraf will connect to the existing stream
+    ## and fail if it does not exist.
+    # external_stream_config = false
```
Comment (srebhan) [inline plugins/outputs/nats/sample.conf]: If I understand this whole PR correctly that flag would prevent the Telegraf plugin from creating the stream but assuming it's already there, doesn't it? If so why not name it as such as your flag name suggests something more... How about

```suggestion
    ## Disable creating the stream but assume the stream is managed externally
    ## and already exists. This will make the plugin fail if the steam does not exist.
    # disable_stream_creation = false
```

File: plugins/outputs/nats/nats_test.go [hunk 6]
```diff
@@ -183,3 +248,30 @@ func TestConfigParsing(t *testing.T) {
...
+	}
+
+	// Add the stream
+	_, err = js.AddStream(&streamConfig)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (srebhan) [inline plugins/outputs/nats/nats_test.go]: How about
```suggestion
func createStream(t *testing.T, server string, cfg *nats.StreamConfig) {
	t.Helper()
	
	// Connect to NATS server
	conn, err := nats.Connect(server)
	require.NoError(t, err)
	
	defer func() {
		require.NoError(t, conn.Drain(), "draining failed")
	}()

	// Create the stream in the JetStream context
	js, err := conn.JetStream()
	require.NoError(t, err)
	_, err = js.AddStream(cfg)
	require.NoError(t, err)
}
```

Of course you would also need to adapt the caller side...

File: plugins/outputs/nats/nats_test.go [hunk 7]
```diff
@@ -119,6 +176,14 @@ func TestConnectAndWriteIntegration(t *testing.T) {
...
+			// If nats cli setup commands are required they need to run
+			// in a nats cli container. The server does not contain the
+			// nats cli tool.
+			if len(tc.externalStream.Name) > 0 {
+				err := createStream(fmt.Sprintf("nats://%s:%s", tc.container.Address, tc.container.Ports[natsServicePort]), tc.externalStream)
+				require.NoError(t, err)
+			}
+
 			server := []string{fmt.Sprintf("nats://%s:%s", tc.container.Address, tc.container.Ports[natsServicePort])}
 			tc.nats.Servers = server
```
Comment (srebhan) [inline plugins/outputs/nats/nats_test.go]: I think you could reuse the address below so how about
```suggestion
			server := "nats://" + tc.container.Address + ":" + tc.container.Ports[natsServicePort]
			
			// Create the stream before starting the plugin to simulate
			// externally managed streams
			if len(tc.externalStream.Name) > 0 {
				createStream(t, server, &tc.externalStream)
			}

			tc.nats.Servers = []string{server}
```

File: plugins/outputs/nats/nats.go [hunk 8]
```diff
@@ -139,11 +140,24 @@ func (n *NATS) Connect() error {
 		if err != nil {
 			return fmt.Errorf("failed to connect to jetstream: %w", err)
 		}
+
+		if n.Jetstream.DisableStreamCreation {
+			stream, err := n.jetstreamClient.Stream(context.Background(), n.Jetstream.Name)
+			if err != nil {
+				if errors.Is(err, nats.ErrStreamNotFound) {
+					return fmt.Errorf("stream %q does not exist and external_stream_config is true", n.Jetstream.Name)
```
Comment (skartikey) [inline plugins/outputs/nats/nats.go]: The error message references `external_stream_config` but the actual config field is `disable_stream_creation`:
```suggestion
					return fmt.Errorf("stream %q does not exist and disable_stream_creation is true", n.Jetstream.Name)
```

File: plugins/outputs/nats/nats.go [hunk 9]
```diff
@@ -139,11 +140,24 @@ func (n *NATS) Connect() error {
...
+				}
+				return fmt.Errorf("failed to get stream info, name: %s, err: %w", n.Jetstream.Name, err)
+			}
+			subjects := stream.CachedInfo().Config.Subjects
+			n.Log.Infof("Connected to existing stream %q with subjects: %v", n.Jetstream.Name, subjects)
```
Comment (skartikey) [inline plugins/outputs/nats/nats.go:153]: **Optional**: Consider adding a warning if the plugin's configured subject isn't in the stream's subject list:
```suggestion
			subjects := stream.CachedInfo().Config.Subjects
			if !choice.Contains(n.Subject, subjects) {
                              n.Log.Warnf("Subject %q is not in stream's configured subjects: %v", n.Subject, subjects)
                          }
                         n.Log.Infof("Connected to existing stream %q with subjects: %v", n.Jetstream.Name, subjects)
```

**Review by skartikey — COMMENTED** [review]
@mgale Nice work! 

A couple of observations:

1. **Great validation logic** - I particularly like how the `streamConfigCompareFunc` proves that the plugin respects external stream configuration rather than its own config values. The contrast between plugin config (`MaxMsgs: 10, MaxConsumers: 100`) vs external config (`MaxMsgs: 10101, MaxConsumers: 6`) clearly demonstrates the feature working as intended.

2. **Minor suggestion**: Consider adding a corner case test for when the external stream exists but doesn't include the plugin's subject:

```go
{
    name: "external stream without plugin subject",
    externalStream: nats.StreamConfig{
        Name:     "my-external-stream", 
        Subjects: []string{"other.subject"}, // Missing "telegraf"
    },
    nats: &NATS{
        Subject: "telegraf",
        Jetstream: &StreamConfig{
            Name: "my-external-stream",
            DisableStreamCreation: true,
        },
    },
    // Should this succeed or fail? Depends on requirements
}
```

@srebhan - what should happen when the external stream exists but doesn't include the plugin's subject? Should the plugin fail with a clear error, or should it succeed and potentially have messages rejected by JetStream at publish time?

File: plugins/outputs/nats/nats.go (hunk 9 above)
Comment (mgale) [reply plugins/outputs/nats/nats.go:153]: Based on my conversation in the NATS Slack (JetStream channel), this kind of validation could be expensive, particularly in larger clusters with many clients. The initial feedback I received suggested that we should simply log publish failures in the plugin’s Write method rather than attempt proactive validation.

I’m not entirely sure how significant the performance impact would be in practice, or how much the validation would actually benefit the broader community.

One option could be to make this behavior configurable, perhaps by adding a parameter like `validate_stream_config` that enables the extra validation when explicitly set.

**Comment by srebhan** [comment]
@skartikey I don't think we should do validations that belong to the server. We cannot know if the subject might be created _after_ we connect so I think we should keep the checks as-is...

File: plugins/outputs/nats/nats.go (hunk 9 above)
Comment (skartikey) [reply plugins/outputs/nats/nats.go:153]: @mgale Thanks for looking into this and sharing the feedback.

Given the potential performance impact, especially in larger clusters, and the uncertainty around the practical benefit, I’d suggest we **don’t add the validation logic** at this point. Keeping the plugin lightweight and focused on logging publish failures seems like the better approach for now.

We can always revisit this later if there’s stronger demand or clearer use cases from the community.

**Review by skartikey — APPROVED** [review]
